core/libp2p/transport/impl/upgrader_impl.cpp

```diff
@@ -57,51 +58,78 @@ namespace libp2p::transport {
         });
   }
 
-  void UpgraderImpl::upgradeToSecure(RawSPtr conn, OnSecuredCallbackFunc cb) {
-    auto initiator = conn->isInitiator();
+  void UpgraderImpl::upgradeToSecureInbound(RawSPtr conn,
+                                            OnSecuredCallbackFunc cb) {
+    protocol_muxer_->selectOneOf(
+        security_protocols_, conn, conn->isInitiator(),
+        [self{shared_from_this()}, cb = std::move(cb),
+         conn](outcome::result<peer::Protocol> proto_res) mutable {
+          if (!proto_res) {
+            return cb(proto_res.error());
+          }
+
+          auto adaptor =
+              findAdaptor(self->security_adaptors_, proto_res.value());
+          if (adaptor == nullptr) {
+            return cb(Error::NO_ADAPTOR_FOUND);
+          }
+
+          BOOST_ASSERT_MSG(!conn->isInitiator(),
```

Comment:
Isn't assert a little too hard? Maybe a special result error is better?

Reply:
it is a dev error if you call `secureOutbound` on inbound connection, so program should indicate that. it is invariant